Lab4/src/ca/mcgill/ecse211/lab4/LightLocalizer.java

```diff
@@ -0,0 +1,300 @@
+    int numlines = 0;
+    double distanceCorr = 0;
+    leftMotor.forward();
+    rightMotor.backward();
+    try {
+      Thread.sleep(1000);
+    } catch (InterruptedException e) {
+      // there is nothing to be done here because it is not
+      // expected that the odometry correction will be
+      // interrupted by another thread
+    }
+    while(true) {
+      correctionStart = System.currentTimeMillis();
+      //fetching the values from the color sensor
+      colorSensorValue.fetchSample(colorSensorData, 0);
+      //getting the value returned from the sensor, and multiply it by 1000 to scale
+      float value = colorSensorData[0]*1000;
+      //computing the derivative at each point
+      float diff = value - oldValue;
+      //storing the current value, to be able to get the derivative on the next iteration
+      oldValue = value;
+      //if the derivative value at a given point is less than -25, this means that a black line is detected
+      if(diff < DIFF_THREASHOLD) {
+        //first line spotted (0 degree line)
+        if(numlines == 0) {
+          //beeps
+          Sound.beep();
+          //calculate the correction needed
+          distanceCorr = Math.sin(odometer.getTheta()*Math.PI/180)*DISTANCE_FROM_CENTER;
+          Nav.turnClockwise();
+          numlines++;
+          //wait 1 second to avoid seeing the same line twice
+          try {
+            Thread.sleep(1000);
+          } catch (InterruptedException e) {
+            // there is nothing to be done here because it is not
+            // expected that the odometry correction will be
+            // interrupted by another thread
+          }
+          //90 degrees line spotted
+        } else if(numlines == 1) {
+          Sound.beep();
+          //correct the error
+          Nav.stopMotors();
+          Nav.goForward();
+          leftMotor.rotate(Nav.convertDistance(distanceCorr),true);
+          rightMotor.rotate(Nav.convertDistance(distanceCorr),false);
+          Nav.turnAnticlockwise();
+          numlines++;
+          //90 degrees line spotted again
+        } else {
+          //as I always spot the beginning of lines, I need to add an offset of half of a black line's angle
+          //to my turn
+          Nav.turnTo(-90-LINE_HALF_ANGLE);
+          Nav.stopMotors();
+          //we are done, reset the odometer to 0
+          odometer.setPosition(new double [] {0.0, 0.0, 0}, new boolean [] {true, true, true});
+          //exit the loop
+          break;
+        }
+      }
+      correctionEnd = System.currentTimeMillis();
+      if (correctionEnd - correctionStart < CORRECTION_PERIOD) {
+        try {
+          Thread.sleep(CORRECTION_PERIOD - (correctionEnd - correctionStart));
+        } catch (InterruptedException e) {
+          // there is nothing to be done here because it is not
+          // expected that the odometry correction will be
+          // interrupted by another thread
+        }
+      }
+    }
+  }
+

```

Comment:
These methods are all very similar and repetitive. Refactor them by extracting the common parts to one common method.